dev/src/bulk-writer.ts

```diff
@@ -73,124 +76,91 @@ const RATE_LIMITER_MULTIPLIER = 1.5;
-  OPEN,
-  READY_TO_SEND,
-  SENT,
-}
-
 /**
  * Used to represent a batch on the BatchQueue.
  *
  * @private
  */
-class BulkCommitBatch {
-  /**
-   * The state of the batch.
-   */
-  state = BatchState.OPEN;
-
+class BulkCommitBatch extends WriteBatch {
+  constructor(firestore: Firestore, readonly maxBatchSize: number) {
+    super(firestore);
+  }
   // The set of document reference paths present in the WriteBatch.
   readonly docPaths = new Set<string>();
 
   // An array of pending write operations. Only contains writes that have not
   // been resolved.
   private pendingOps: Array<Deferred<BatchWriteResult>> = [];
 
-  private readonly backoff: ExponentialBackoff;
-
-  constructor(
-    private readonly firestore: Firestore,
-    private writeBatch: WriteBatch,
-    private readonly maxBatchSize: number
-  ) {
-    this.backoff = new ExponentialBackoff();
+  _has(documentRef: firestore.DocumentReference): boolean {
+    return this.docPaths.has(documentRef.path);
   }
 
-  /**
-   * The number of writes in this batch.
-   */
-  get opCount(): number {
-    return this.pendingOps.length;
+  _markReadyToSend(): void {
+    if (this.state === BatchState.OPEN) {
+      this.state = BatchState.READY_TO_SEND;
+    }
   }
 
-  /**
-   * Adds a `create` operation to the WriteBatch. Returns a promise that
-   * resolves with the result of the write.
-   */
-  create<T>(
-    documentRef: firestore.DocumentReference<T>,
-    data: T
-  ): Promise<WriteResult> {
-    this.writeBatch.create(documentRef, data);
-    return this.processLastOperation(documentRef);
+  _isOpen(): boolean {
+    return this.state === BatchState.OPEN;
   }
 
-  /**
-   * Adds a `delete` operation to the WriteBatch. Returns a promise that
-   * resolves with the sentinel value (Timestamp(0)) for the delete operation.
-   */
-  delete<T>(
-    documentRef: firestore.DocumentReference<T>,
-    precondition?: firestore.Precondition
-  ): Promise<WriteResult> {
-    this.writeBatch.delete(documentRef, precondition);
-    return this.processLastOperation(documentRef);
+  _isReadyToSend(): boolean {
+    return this.state === BatchState.READY_TO_SEND;
   }
 
-  set<T>(
-    documentRef: firestore.DocumentReference<T>,
-    data: Partial<T>,
-    options: firestore.SetOptions
-  ): Promise<WriteResult>;
-  set<T>(
-    documentRef: firestore.DocumentReference<T>,
-    data: T
-  ): Promise<WriteResult>;
-  set<T>(
-    documentRef: firestore.DocumentReference<T>,
-    data: T | Partial<T>,
-    options?: firestore.SetOptions
-  ): Promise<WriteResult>;
-  /**
-   * Adds a `set` operation to the WriteBatch. Returns a promise that
-   * resolves with the result of the write.
-   */
-  set<T>(
-    documentRef: firestore.DocumentReference<T>,
-    data: T | Partial<T>,
-    options?: firestore.SetOptions
-  ): Promise<WriteResult> {
-    this.writeBatch.set(documentRef, data, options);
-    return this.processLastOperation(documentRef);
-  }
+  async bulkCommit(): Promise<void> {
```

Comment:
s/bulkCommit/_bulkCommit

But really should this maybe just be `commit` and overwrite the `commit` method in the base class? That way, it hides the existing implementation which we would never want to use.

Reply:
Why are we adding the underscore for this method, but not the others?

I originally wanted to just override the base class, but `WriteBatch.commit()` returns `Promise<WriteResult[]>`, which doesn't match the signature here.

Reply:
I thought this was a public class. It's not, so we don't need to add it. Thanks for pointing out the different return types. I think it might be best to leave as is then.